### PR TITLE
Edit and delete user

### PR DESCRIPTION
**Pull request by sherryhli**
* `preferredName` and `deleted` added to user schema
* User endpoint to delete user from both MongoDB and Firebase (soft delete from MongoDB, actual removal from Firebase)
* Edit and delete user functionality from user management table

image
image

### CODE REVIEW

File: src/Reducers/UserFormStateReducer.js [hunk 1]
```diff
@@ -26,6 +26,8 @@ export function userFormStateReducer(formState, action) {
           ? draftFormState.programs.delete(action.program)
           : draftFormState.programs.add(action.program);
         break;
+      default:
```
Comment (sherryhli) [inline src/Reducers/UserFormStateReducer.js:29]: I know I previously asked for this to be removed but turns out React warns against missing defaults 😅

File: src/Components/Admin/DialogWrappers.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  display: inline-block;
+  left: 50%;
+  margin: auto;
+  padding: ${padding};
+  position: fixed;
+  top: 50%;
+  width: 400px;
+  z-index: 1000;
+
+  -moz-transform: translate(-50%, -50%);
```
Comment (GodGreg) [inline src/Components/Admin/DialogWrappers.js:15]: I don't think this is needed. Not critical but we should probably clean up this css a bit

File: backend/routes/users.js [hunk 3]
```diff
@@ -47,6 +48,29 @@ router.post("/", function(req, res) {
...
+        res.status(500).send(err);
+      } else {
+        deleteFirebaseUser(req.params.userId)
+          .then(() => {
+            res.status(204).send();
+          })
+          .catch((err) => {
+            console.log(`User with UID = ${req.params.userId}
+              was marked deleted in MongoDB but not removed from Firebase, failed due to: ${err}`);
+            // using status code 202 (accepted) to represent partial success
```
Comment (GodGreg) [inline backend/routes/users.js:66]: Ok. Good idea

File: src/Components/Admin/UserManagement.jsx [hunk 4]
```diff
@@ -26,36 +26,56 @@ const Header = styled.div`
...
 
 function UserManagement() {
   const [users, setUsers] = useState(null);
 
   useEffect(() => {
+    //Option 2: Define an async function to call at the end of useEffect
     // Define an async function since callback must be synchronous
     async function getUsers() {
       // Fetch the users from the backend
       const fetched = await GET.getAllUsersAPI();
```
Comment (GodGreg) [inline src/Components/Admin/UserManagement.jsx:76]: Looks like these comments are already implemented?

File: src/Reducers/UserFormStateReducer.js (hunk 1 above)
Comment (GodGreg) [reply src/Reducers/UserFormStateReducer.js:29]: Yeah i added it back in during the merge :)

File: src/Components/Admin/DialogWrappers.js (hunk 2 above)
Comment (sherryhli) [reply src/Components/Admin/DialogWrappers.js:15]: No I think all the translations are needed to position the dialog properly, we'll need to keep them until we transition to drawers

Comment (GodGreg) [reply src/Components/Admin/DialogWrappers.js:15]: Sounds good. Just never used it before so I didn't know what it did ¯\_(ツ)_/¯

File: src/Components/Admin/UserManagement.jsx [hunk 5]
```diff
@@ -85,4 +103,27 @@ function UserManagement() {
...
+//         name: "",
+//         email: item.email,
+//         role: "User",
+//         programs: programs
+//       };
+//       //2 FACTOR AUTH. CONTACT GREG BEFORE UNCOMMENTING
+//       //UPDATE.updateUserAPI(user)
+//     });
+//   });
+// }
```
[the remark this answers is not in the document]
Comment (GodGreg) [reply src/Components/Admin/UserManagement.jsx:127]: Possible future use case. I added it back in from a merge.

File: src/Components/Application/Application.jsx [hunk 6]
```diff
@@ -119,11 +119,13 @@ function Application({ applications, newReview, history, match, user }) {
 
   return (
     <PageWrapper>
-      <FlowSelector>
-        <button>1. Letter of Interest</button>
-        <button disabled>2. Full Application</button>
-      </FlowSelector>
       <BodyWrapper>
+        {/* //This should be in the body wrapper
```
[the remark this answers is not in the document]
Comment (GodGreg) [reply src/Components/Application/Application.jsx:123]: Yes. It is currently not being used and was just hidden behind the header.

May be used for the new navbar progress

File: src/Components/Admin/UserManagement.jsx [hunk 7]
```diff
@@ -26,36 +26,54 @@ const Header = styled.div`
       margin-right: auto;
     }
     .button-container {
-      display: inline-block;
+      display: inline-block; //This may need to be flex (changed during a merge conflict)
```
[the remark this answers is not in the document]
Comment (GodGreg) [reply src/Components/Admin/UserManagement.jsx:29]: inline-block seems fine - although I didn't extensively test this with different screen sizes.

[the remark this answers is not in the document]
Comment (sherryhli) [reply src/Components/Admin/UserManagement.jsx]: inline-block, also took a look through our commit history and looks like it was inline-block to begin with?

**Pull request merged**